### ISSUE TITLE
Build error ‘channels’ may be used uninitialized in this function

### ISSUE THREAD

**Issue by yuvadm** [fdb/frequensea#2]
```
[ 82%] Building C object CMakeFiles/frequensea.dir/src/nim.c.o
/home/yuval/dev/archlinux-packages/aur/frequensea-git/src/frequensea/src/nim.c: In function ‘nim_png_write’:
/home/yuval/dev/archlinux-packages/aur/frequensea-git/src/frequensea/src/nim.c:40:9: error: ‘channels’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
     int channels;
         ^
cc1: all warnings being treated as errors
CMakeFiles/frequensea.dir/build.make:123: recipe for target 'CMakeFiles/frequensea.dir/src/nim.c.o' failed
make[2]: *** [CMakeFiles/frequensea.dir/src/nim.c.o] Error 1
CMakeFiles/Makefile2:60: recipe for target 'CMakeFiles/frequensea.dir/all' failed
make[1]: *** [CMakeFiles/frequensea.dir/all] Error 2
Makefile:76: recipe for target 'all' failed
make: *** [all] Error 2
==> ERROR: A failure occurred in build().
```

**Comment by fdb** [comment]
Hi Yuval,

I’m kinda busy today, but I’ll try to get a Linux box so I can test this properly. PRs are also welcome :-)


Best,

F

On 13 Mar 2015 at 11:51:40, Yuval Adam (notifications@github.com) wrote:

[ 82%] Building C object CMakeFiles/frequensea.dir/src/nim.c.o
/home/yuval/dev/archlinux-packages/aur/frequensea-git/src/frequensea/src/nim.c: In function ‘nim_png_write’:
/home/yuval/dev/archlinux-packages/aur/frequensea-git/src/frequensea/src/nim.c:40:9: error: ‘channels’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
     int channels;
         ^
cc1: all warnings being treated as errors
CMakeFiles/frequensea.dir/build.make:123: recipe for target 'CMakeFiles/frequensea.dir/src/nim.c.o' failed
make[2]: *** [CMakeFiles/frequensea.dir/src/nim.c.o] Error 1
CMakeFiles/Makefile2:60: recipe for target 'CMakeFiles/frequensea.dir/all' failed
make[1]: *** [CMakeFiles/frequensea.dir/all] Error 2
Makefile:76: recipe for target 'all' failed
make: *** [all] Error 2
==> ERROR: A failure occurred in build().

**Comment by yuvadm** [comment]
I see some other errors popping up as well. I'll try to send some patches your way :)

**Comment by altercation** [comment]
Were there any patches for this issue? I'm hitting the same problem under Arch.

**Comment by fdb** [comment]
@yuvadm has posted a PR which solves some of the first errors. I still have to find the time to get everything working. 

I'll try to do this next week.

**Comment by fdb** [comment]
Using the current master, Frequensea should build (although not without warnings, most of them caused by including the GLFW source).

However, on my testing machine (Chromebook with Ubuntu 14.04) only the `../lua/empty.lua` script works correctly. This has probably to do with support for OpenGL ES. Somebody with a OpenGL Desktop system, please test this.

**Comment by yuvadm** [comment]
Getting:

```
[ 55%] Building C object CMakeFiles/frequensea.dir/src/nosc.c.o
/home/yuval/dev/archlinux-packages/aur/frequensea-git/src/frequensea/src/nosc.c: In function ‘_nosc_server_start’:
/home/yuval/dev/archlinux-packages/aur/frequensea-git/src/frequensea/src/nosc.c:259:17: warning: implicit declaration of function ‘nanosleep’ [-Wimplicit-function-declaration]
                 nanosleep((struct timespec[]){{0, 1 * MILLIS_TO_NANOS}}, NULL);
                 ^
/home/yuval/dev/archlinux-packages/aur/frequensea-git/src/frequensea/src/nosc.c:259:35: error: array type has incomplete element type
                 nanosleep((struct timespec[]){{0, 1 * MILLIS_TO_NANOS}}, NULL);
                                   ^
CMakeFiles/frequensea.dir/build.make:169: recipe for target 'CMakeFiles/frequensea.dir/src/nosc.c.o' failed
make[2]: *** [CMakeFiles/frequensea.dir/src/nosc.c.o] Error 1
CMakeFiles/Makefile2:61: recipe for target 'CMakeFiles/frequensea.dir/all' failed
make[1]: *** [CMakeFiles/frequensea.dir/all] Error 2
Makefile:116: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by altercation** [comment]
Confirming identical result as @yuvadm here (building on arch surface pro 3)

**Comment by yuvadm** [comment]
Per http://stackoverflow.com/questions/14415465/why-do-i-get-in-my-timespec-array-the-error-array-type-has-incomplete-element-t this should fix the problem:

```c
#if __STDC_VERSION__ >= 199901L
#define _XOPEN_SOURCE 600
#else
#define _XOPEN_SOURCE 500
#endif /* __STDC_VERSION__ */
#include <time.h>
```

However even the warning about implicit use of `nanosleep` still exists. Is there some other `time.h` in use here?

**Comment by fdb** [comment]
I found out that I already fixed the issue in nrf.c -- so I moved the implementation to the `nut` utility module. The code should build now.

Since the code also includes glfw as externals, it tries to build GLFW's examples and fails. However, the `frequensea` binary should still build.

**Comment by yuvadm** [comment]
@fdb awesome! Is there any way to disable the GLFW examples? Locally it's fine, but I can't package something that fails a build. If it's not required I'd just like to disable it.

**Comment by yuvadm** [comment]
I'm also getting this:

```bash
$ ./frequensea ../lua/empty.lua 
GLFW ERROR 65543: GLX: Failed to create context: GLXBadFBConfig
```

Happens on all the scripts.

**Comment by fdb** [comment]
Hmm have you been able to run other OpenGL stuff? Do you have other apps that use GLFW?

Maybe I should switch back to the GLFW packaged release. I believe Arch has a recent GLFW 3 package, no?

**Comment by yuvadm** [comment]
@fdb yes, there's no reason to bundle GLFW, any proper distribution should have it packaged separately. If you do decide to keep it, at least provide a flag for package maintainers to disable it.

As for other OpenGL stuff, I'll test some more and see what I come up with.

**Comment by fdb** [comment]
@yuvadm you're right, of course. I did this because Ubuntu 14.04 doesn't have GLFW3 in its repository, only GLFW2.

**Comment by fdb** [comment]
I removed GLFW3 from source and uses the one that should be pre-installed. I had to use `pkg-config` to include all relevant X11 libraries.

**Comment by fdb** [comment]
As for the GLXBadFBConfig, it's because you're using OpenGL 2.1 (Mesa) drivers. We require at least OpenGL 3.2 (See [nwm.c](https://github.com/fdb/frequensea/blob/f4497c2f98b63972a79be5b0c39735812a7dc1e0/src/nwm.c#L25)).

Depending on your video card, proprietary drivers might include a more recent version of OpenGL.

**Comment by yuvadm** [comment]
@fdb build seems to finish successfully, thanks for the updates! I'll go ahead and test on a machine with better OpenGL support.

**Comment by altercation** [comment]
Package builds under arch but I get an immediate segfault when running frequensea on any of the test lua files. I'll investigate further as well. OpenGL supported fine on this machine so there's something else happening. Any tips on diagnosing or information I can provide?

**Comment by fdb** [comment]
Try running using gdb:

    gdb --args ./frequensea ../lua/static.lua
    > run
    > backtrace full

Thanks for helping me debug this issue!
